silence/silence.go

```diff
@@ -671,252 +663,149 @@ func (s *Silences) query(q *query, now time.Time) ([]*pb.Silence, error) {
+			e.Silence.CreatedBy = e.Silence.Comments[0].Author
+			e.Silence.Comments = nil
 		}
+		st[e.Silence.Id] = e
 	}
-
-	s.st.data = st.data
+	s.mtx.Lock()
+	s.st = st
+	s.mtx.Unlock()
 
 	return nil
 }
 
 // Snapshot writes the full internal state into the writer and returns the number of bytes
 // written.
-func (s *Silences) Snapshot(w io.Writer) (int, error) {
+func (s *Silences) Snapshot(w io.Writer) (int64, error) {
 	start := time.Now()
 	defer func() { s.metrics.snapshotDuration.Observe(time.Since(start).Seconds()) }()
 
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
 
-	var n int
-	for _, s := range s.st.data {
-		m, err := pbutil.WriteDelimited(w, s)
-		if err != nil {
-			return n + m, err
-		}
-		n += m
+	b, err := s.st.MarshalBinary()
+	if err != nil {
+		return 0, err
 	}
-	return n, nil
-}
 
-type gossiper struct {
-	*Silences
+	return io.Copy(w, bytes.NewReader(b))
 }
 
-// Gossip implements the mesh.Gossiper interface.
-func (g gossiper) Gossip() mesh.GossipData {
-	g.mtx.Lock()
-	defer g.mtx.Unlock()
+// MarshalBinary serializes all contents of the notification log.
+func (s *Silences) MarshalBinary() ([]byte, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
 
-	return g.st.clone()
+	return s.st.MarshalBinary()
 }
 
-// OnGossip implements the mesh.Gossiper interface.
-func (g gossiper) OnGossip(msg []byte) (mesh.GossipData, error) {
-	gd, err := decodeGossipData(msg)
+func (s *Silences) Merge(b []byte) error {
+	st, err := decodeState(bytes.NewReader(b))
 	if err != nil {
-		return nil, err
+		return err
 	}
-	g.mtx.Lock()
-	defer g.mtx.Unlock()
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
 
-	if delta := g.st.mergeDelta(gd); len(delta.data) > 0 {
-		return delta, nil
+	for _, e := range st {
+		s.st.merge(e)
 	}
-	return nil, nil
+	return nil
 }
 
-// OnGossipBroadcast implements the mesh.Gossiper interface.
-func (g gossiper) OnGossipBroadcast(src mesh.PeerName, msg []byte) (mesh.GossipData, error) {
-	gd, err := decodeGossipData(msg)
-	if err != nil {
-		return nil, err
+func (s *Silences) MergeSingle(b []byte) error {
+	var e pb.MeshSilence
+	if err := proto.Unmarshal(b, &e); err != nil {
+		return err
 	}
-	g.mtx.Lock()
-	defer g.mtx.Unlock()
-
-	return g.st.mergeDelta(gd), nil
-}
-
-// OnGossipUnicast implements the mesh.Gossiper interface.
-// It always panics.
-func (g gossiper) OnGossipUnicast(src mesh.PeerName, msg []byte) error {
-	panic("not implemented")
+	s.mtx.Lock()
+	s.st.merge(&e)
+	s.mtx.Unlock()
+	return nil
 }
 
-type silenceMap map[string]*pb.MeshSilence
-
-type gossipData struct {
-	data silenceMap
-	mtx  sync.RWMutex
+func (s *Silences) SetBroadcast(f func([]byte)) {
+	s.mtx.Lock()
+	s.broadcast = f
+	s.mtx.Unlock()
 }
 
-var _ mesh.GossipData = &gossipData{}
+type state map[string]*pb.MeshSilence
 
-func newGossipData() *gossipData {
-	return &gossipData{
-		data: silenceMap{},
+func (s state) clone() state {
+	c := make(state, len(s))
+	for k, v := range s {
```

Comment:
I can't tell from this PR, but if the caller is not protecting this call, then it seems like this needs a read lock

Reply:
It's just some sugar for a map – so yea, the caller has to lock.